File: cpp/388/D.cpp
#include <iostream>
#include <vector>
using namespace std;

int main() {
    int N;
    cin >> N;
    vector<int> A(N);

    for (int i = 0; i < N; ++i) {
        cin >> A[i];
    }

    for (int i = 0; i < N; ++i) {
        int to_dis;
        if (A[i] < N - 1 - i) {
            to_dis = A[i];
        } else {
            to_dis = N - 1 - i;
        }
        A[i] -= to_dis;
        for (int j = i+1; j < i+to_dis+1; ++j) {
            A[j] += 1; 
        }
    }

    for (int i = 0; i < N; ++i) {
        cout << A[i];
        if (i < N - 1) cout << " ";
    }
    cout << endl;

    return 0;
}

File: python/388/D.py
N = int(input())
A = [int(num) for num in input().split()]

temp = []

for i in range(N):
    to_add,end = 0,0
    if len(temp) > 0:
        to_add, end = temp[len(temp) - 1]
        if i > end:
            temp.pop()
    to_dis = min(A[i]+to_add, N-1-i)
    A[i] += to_add
    A[i] -= to_dis
    temp.append((i+1, i+to_dis))
    # print(i, to_dis, A)

print(' '.join(map(str, A)))
